kafka-client/src/main/java/dev/responsive/kafka/api/ResponsiveGlobalConsumer.java

```diff
@@ -0,0 +1,398 @@
+ *   can properly use the committed offsets.</li>
+ *
+ *   <li>It allows looking up {@link #position(TopicPartition)} for partitions
+ *   that this consumer does not own by proxying the Admin for those partitions.</li>
+ *
+ *   <li>It returns {@link ConsumerRecords} from {@link #poll(Duration)} that will
+ *   return all records when calling {@link ConsumerRecords#records(TopicPartition)}
+ *   instead of only the requested partition. This is necessary because of the way that
+ *   the {@link org.apache.kafka.streams.processor.internals.GlobalStateManagerImpl}
+ *   handles restore. Specifically, it will loop partition-by-partition and update
+ *   the store for those partitions. The issue with this is that if we use {@code subscribe}
+ *   in place of {@code assign} we may get events from partitions that were not
+ *   specified in the {@code assign} call. These records will be dropped.</li>
+ *
+ *   <li>It commits before every call to {@link #poll(Duration)} because
+ *   otherwise the Streams implementation for global stores will never commit. We
+ *   should consider changing this to every {@code N} polls, but this would be
+ *   more difficult to test (we could make it configurable).</li>
+ * </ol>
+ *
+ * <p>This class breaks a lot of abstraction barriers, but allows us to
+ * support remote {@link org.apache.kafka.streams.kstream.GlobalKTable}s without
+ * forking Kafka Streams.</p>
+ */
+public class ResponsiveGlobalConsumer implements Consumer<byte[], byte[]> {
+
+  private final Consumer<byte[], byte[]> delegate;
+  private final Integer defaultApiTimeoutMs;
+  private final Admin admin;
+
+  public ResponsiveGlobalConsumer(
+      final Map<String, Object> config,
+      final Consumer<byte[], byte[]> delegate,
+      final Admin admin
+  ) {
+    final ConsumerConfig consumerConfig = new ConsumerConfig(config);
+
+    this.delegate = delegate;
+    this.defaultApiTimeoutMs = consumerConfig.getInt(ConsumerConfig.DEFAULT_API_TIMEOUT_MS_CONFIG);
+    this.admin = admin;
+  }
+
+  @Override
+  public Set<TopicPartition> assignment() {
+    return delegate.assignment();
+  }
+
+  @Override
+  public Set<String> subscription() {
+    return delegate.subscription();
+  }
+
+  @Override
+  public void subscribe(final Collection<String> topics, final ConsumerRebalanceListener listener) {
+    delegate.subscribe(topics, listener);
+  }
+
+  @Override
+  public void subscribe(final Collection<String> topics) {
+    delegate.subscribe(topics);
+  }
+
+  @Override
+  public void subscribe(final Pattern pattern, final ConsumerRebalanceListener listener) {
+    delegate.subscribe(pattern, listener);
+  }
+
+  @Override
+  public void subscribe(final Pattern pattern) {
+    delegate.subscribe(pattern);
+  }
+
+  @Override
+  public void unsubscribe() {
+    delegate.unsubscribe();
+  }
+
+  @Override
+  public void assign(final Collection<TopicPartition> partitions) {
+    subscribe(
```

Comment:
I'm a little worried about unintended side effects of repeatedly calling `subscribe` on an ever-changing set of topics...For example, a consumer group's committed offsets will eventually be deleted for topics that no one has subscribed to. Not sure off the top of my head exactly what determines the retention semantics there, I can look into it I guess.

It would be great if we could just store the set of all topics we've seen so far and stay subscribed to all of them, but of course this is thwarted by the GlobalStateManager's ridiculous restoration loop which restores one store at a time. Unfortunately this also makes it all the more likely for us to lose one topic's offsets while restoring state from another, since it seems to literally go through each entire global store, one by one.

Hopefully they only have one global table? I didn't even think to ask them... :/

Reply:
this should happen pretty infrequently right? just on restore and then once when the global thread starts up normally per stream app. if we really wanted to avoid some overhead we could check whether or not the topics are already subscribed to and then not make the call to `subscribe`?